### ISSUE TITLE
Onboarding process with mmol/l as native unit

### ISSUE THREAD

**Issue by hpc4fun** [LoopKit/Loop#1515]
**Describe the bug**
When doing the new loop-onboarding procedure I noticed that the current sequence of events gives rise to a issue for users with mmol/l as there native unit. The current onboarding works fine for mg/dl but is less user-friendly to mmol/l users. Allowance of access to HK (which could  tell loop about the prefered unit) is not the first question in the on-boarding process so loop will assume mg/dl. Moreover, there is no other ovious way that one can switch between mmol/l and mg/dl in the sequence of on-boarding questions so one will have to do all initial settings in mg/dl and only after the HK access approval, all the settings will then be converted back to mmol/l so this appears solely as an issue with the sequence of the events and not with missing functionality per see. Maybe one could start with the HK approval (like in current release version) or alternatively give users the option to specify the unit for BG as the very first step in the on-boarding process and that would be it.

**To Reproduce**
Follow the onboarding process with the assumption that you default unit for all settings is mmol/l.

**Expected behavior**
I expect that users should be able to follow the on-boarding process using their native unit for BG.

**Loop Version**
 - Version Number: 2.3.0 (57),  [e.g. 1.9.2]
 - Repo: LoopKit/LoopWorkspace (453ee6a)

**Comment by hpc4fun** [comment]
Second thought on a fix: I would actually prefer that the unit was specified directly to loop (instead of indirectly via HK) and also stored statically. The reason for this is that each time one updates IOS the unit info is lost after the first reboot and it starts up with the default unit (mg/dl) which for some users is the wrong unit and they will have to terminate the app and have it reload the unit from the HK. I guess this glitch could be prevented if the unit was explicitly specified and stored statically.

**Comment by ps2** [comment]
Let's keep this ticket focused on fixing the original issue description. With the current strategy of Loop following your HealthKit preference, we can, and should, ask for HK permissions before asking for glucose values.

There are at least two other items in the follow on comment:

1. A bug or some behavior (first I've heard of it) of units being reset after ios upgrade.  Is this an Apple Health reset of units that you're talking about? Or just Loop? or? Could you open a new issue describing more details about this? Sounds like if we keep the current strategy we should do we can to work better with this behavior.
2. A feature request to allow Loop to operate independently of Apple Health preferred units. Assuming we can fix the onboarding issue, and the unit reset issue. Is there still a use case where you'd want Apple Health and Loop to have different units for glucose? If so, please open another ticket describing why this would be useful.

Thanks!

**Comment by 4gra** [comment]
I too have seen this behaviour following a (non-update) phone reboot: the first launch of Loop shows mg/dl; subsequent kill & relaunch shows mmol/l. However it seems intermittent and I haven't managed to corroborate on a quick test.

(For what it's worth I think the bigger picture is that it's wrong to have an assumption of a default unit, given the chance for error if the user doesn't spot the change. In my opinion, if positive confirmation hasn't been received of native units, Loop shouldn't operate, or should present a warning anyway (since it's a display rather than operational issue). In this way a whole class of errors could be detected if not avoided for mmol/l users.)

**Comment by dabear** [comment]
> A bug or some behavior (first I've heard of it) of units being reset after ios upgrade. Is this an Apple Health reset of units that you're talking about? 

This is most likely due to bluetooth state restore. I've added a notification that gets triggered on staterestore in my local build. Everytime that happens, units gets reset to mgdl instead of my preference which is mmol. I've never seen this behavior without state restoration also happening.

**Comment by hpc4fun** [comment]
> Let's keep this ticket focused on fixing the original issue description. With the current strategy of Loop following your Apple Health preference, we can, and should, ask for HK permissions before asking for glucose values.
> 
> There are at least two other items in the follow on comment:
> 
> 1. A bug or some behavior (first I've heard of it) of units being reset after ios upgrade.  Is this an Apple Health reset of units that you're talking about? Or just Loop? or? Could you open a new issue describing more details about this? Sounds like if we keep the current strategy we should do we can to work better with this behavior.
> 2. A feature request to allow Loop to operate independently of Apple Health preferred units. Assuming we can fix the onboarding issue, and the unit reset issue. Is there still a use case where you'd want Apple Health and Loop to have different units for glucose? If so, please open another ticket describing why this would be useful.
> 
> Thanks!

1) This is just Loop. On rare occasions (say few times a year) you may see loop suddenly operate in mg/dl instead of mmol/l settings. This is *NOT* an issue introduced in dev but one that have persisted in Loop as long as I recall. It's not a big issue and the work-around is simply to close and reopen loop and it will pick up the proper unit from AH and continue in mmol/l mode. Having said that, the only way that I'm aware off that will reproduce this issue is during an IOS upgrade. And I agree, this was a side-comment to the real problem reported here which seems to be solved by sha 8d9072c in loopworkspace. Thanks!

2) I will keep an eye on whether or not the issue with change of units during IOS updates persists and report back if that's still the case after the recent updates on dev.

**Comment by ps2** [comment]
I am hoping that the changes pushed to dev will also fix the "reset". But please report if the reset is seen again, in a new ticket, preferrably.
